javalib/src/main/scala/java/nio/file/attribute/FileTime.scala

```diff
@@ -21,25 +23,31 @@ final class FileTime private (private val epochMillis: Long,
     }
 
   override def hashCode(): Int =
-    epochMillis.## + nanos.##
+    value.## + unit.##
 
-  def to(unit: TimeUnit): Long =
-    unit.convert(toMillis(), TimeUnit.MILLISECONDS)
+  def to(unit: TimeUnit): Long = unit.convert(value, this.unit)
 
-  def toMillis(): Long = epochMillis
+  def toMillis(): Long = to(TimeUnit.MILLISECONDS)
 
-  def toInstant(): Instant = Instant.ofEpochMilli(epochMillis).plusNanos(nanos)
+  def toInstant(): Instant = {
+    val asNanos = to(TimeUnit.NANOSECONDS)
+    val seconds = to(TimeUnit.SECONDS)
+    Instant.ofEpochSecond(
+      seconds,
+      asNanos - TimeUnit.NANOSECONDS.convert(seconds, TimeUnit.SECONDS))
+  }
 
-  override def toString(): String = s"FileTime($epochMillis, $nanos)"
+  override def toString(): String = s"FileTime($value, $unit)"
 }
 
 object FileTime {
-  def from(value: Long, unit: TimeUnit): FileTime =
-    fromMillis(TimeUnit.MILLISECONDS.convert(value, unit))
+  def from(value: Long, unit: TimeUnit): FileTime = new FileTime(value, unit)
 
-  def fromMillis(value: Long): FileTime =
-    new FileTime(value, 0)
+  def fromMillis(value: Long): FileTime = from(value, TimeUnit.MILLISECONDS)
 
-  def from(instant: Instant): FileTime =
-    new FileTime(instant.toEpochMilli, instant.getNano)
+  def from(instant: Instant): FileTime = {
+    val secondsAsNanos =
+      TimeUnit.NANOSECONDS.convert(instant.getEpochSecond, TimeUnit.SECONDS)
+    new FileTime(secondsAsNanos + instant.getNano, TimeUnit.NANOSECONDS)
```

Comment:
This will not work for large `Instant`s that cannot be represented in a Long value of nanoseconds. We must start with the Seconds, and only keep more precision (in milliseconds or nanoseconds) if the seconds are <= `Long.MaxValue / 1000` (or `1000000` for nanoseconds).

Reply:
Guess it's outdated, as now, after conversion to (epochDays, dayNanos) we are having a greater range than Instant. Now we need to make sure conversion toInstant does not overexcites Instant limits. To check this additional tests were added to `FileTimeExtTest`